### ISSUE TITLE
accessing elements of degree of freedom arrays in balance law solver

### ISSUE THREAD

**Issue by jkozdon** [climate-machine/CLIMA#245]
After some discussion with @charleskawczynski, @asraero and @trontrytel, we decided that it would be very useful to be able to declare internal model variables that could be accessed via `getfield` overloading instead of e.g. `ρ == Q[_ρ]`.

Maybe something like:
```
@variables MyModel begin
    @constant begin
        # time-constant variables, e.g. coordinates, reference values, etc.
        # currently called "aux" 
        x,y,z
    end

    @diagnostic begin
         # currently computed via "preflux"
         ql
     end

     @prognostic begin
         # state variables
         U,V,W,ρ,Q,E
      end
end
```

It would also be useful to be able to create compound data structures (e.g. structs or static arrays) from combinations of these. An example would be getting a static array of velocities.

_Originally posted by @simonbyrne in https://github.com/climate-machine/CLIMA/pull/233#issuecomment-495404115_

**Comment by jkozdon** [comment]
Can you give a desugared example? The macro should be implementable, but I am worried about GPU compatibility. Are these static variables for the entire Grid, for each element, or for each nodal point? 

What we might be able to do, is to encapsulate the local view into the global data and have the BalanceLaw solver create a local model instance.

I explicitly excluded models having state in the first draft, since I don't quite know how that fits into what the solver needs to do. 

_Originally posted by @vchuravy  in https://github.com/climate-machine/CLIMA/pull/233#issuecomment-495411957

**Comment by jkozdon** [comment]
Something to keep in mind if someone tries to do this:

- Right now we grab all of the states, aux state, and viscous state to pass to the code injection functions. I imagine we'll likely have to move away from this for performance reason and only grab values out of the global arrays the kernel actually needs.
- When we have O(100) tracers, there will be some sub-grouping that has to occur.

Whatever framework is chosen, will need to be flexible enough to support this. The introduction of the `DGModel` type should help here though.

Also, if this were done the above sub-selection would be seen more cleanly by the user possibly.

Finally, we may want to support functions as well. For instance, `Q.U` gives the momentum (`ρu`) and `Q.u` gives the velocity (`U/ρ`). You could even imagine support `Q.u⃗` giving the vector `(u, v, w)`

**Comment by simonbyrne** [comment]
Basically the idea is that the macros would generate the necessary `getproperty`/`setproperty!` overloads, so that any such accesses will be inlined and resolved to regular array indexing at compile time. 

I'm not sure how this fits into your question about the model state, but we can discuss this.

> * Right now we grab all of the states, aux state, and viscous state to pass to the code injection functions. I imagine we'll likely have to move away from this for performance reason and only grab values out of the global arrays the kernel actually needs.

This should help with your first point, since users won't need to splat the array to symbols at the start of their code

> * When we have O(100) tracers, there will be some sub-grouping that has to occur.

Is this to avoid the register spilling? I don't have a good solution for this. The only thing I can think of at the moment is providing a framework for users to do such a grouping. This might be a @vchuravy problem.

> Finally, we may want to support functions as well. For instance, `Q.U` gives the momentum (`ρu`) and `Q.u` gives the velocity (`U/ρ`). You could even imagine support `Q.u⃗` giving the vector `(u, v, w)`

I was thinking something along those lines as well.

**Comment by jkozdon** [comment]
> Basically the idea is that the macros would generate the necessary `getproperty`/`setproperty!` overloads, so that any such accesses will be inlined and resolved to regular array indexing at compile time.

What types would these be overloaded for though? Everything is coming through as an `MArray`?

> This should help with your first point, since users won't need to splat the array to symbols at the start of their code

The balance law solver is still reading these out from the global arrays though. That's what we'll need to avoid possibly at some point.

> Is this to avoid the register spilling? I don't have a good solution for this. The only thing I can think of at the moment is providing a framework for users to do such a grouping.

Yes.

I imagine at some point the user will split the state into an advected state and nonlinear state or something like this. In the nonlinear state we'll need to know what states the user needs to define the  nonlinear flux and the user will have to provide a function or something to give the velocity field... I don't think that we worry about this now, but at some point we will need to.

If the tracers are decoupled from one another we can probably handle the grouping at the balance law solver level if we know which states are advected. This will probably require some iteration once we have a test case it is needed for.

**Comment by jkozdon** [comment]
Overall the new `DGModel` framework should make this all less cumbersome for the user I think.

**Comment by simonbyrne** [comment]
> What types would these be overloaded for though? Everything is coming through as an `MArray`?

We would have a wrapper type for the various arrays (state/aux/etc) which would be parameterised by the model.

> The balance law solver is still reading these out from the global arrays though.

Well, we could possibly avoid the intermediate `MArray` and access the main state array directly?

**Comment by vchuravy** [comment]
So my thinking is that we have a `LocalModel` type that we can overload, the constructor would receive the  data from the balance-law solver as an MArray (or otherwise) and we can then define the getproperty on that type.

We might also be able to use a `view(GlobalData, ...)` instead of a MArray, but in practice we will want to be careful to minimize off-chip data transfers. 

I think the best way forward would be to take a model, and manually translate it into the "ideal" form.

cc: @glwagner who expressed an interest in thinking about this.

**Comment by simonbyrne** [comment]
To follow up on this, I was thinking that the flux here:
https://github.com/climate-machine/CLIMA/blob/94268424f9371f4fc5c32bfa8e7af319c9235814/test/DGmethods/compressible_Navier_Stokes/mms_bc.jl#L66-L85
could be written as
```
flux.ρ  = state.ρ𝐮
flux.ρ𝐮 = state.𝐮 * state.ρ𝐮' + P*I - viscstate.T
flux.E  = state.𝐮 .* (state.E + P) - viscstate.T * state.𝐮
```
where:
- `state.ρ𝐮` is the vector of `[Q[_U], Q[_V], Q[_W]]`
- `state.𝐮` is the vector of `[Q[_U], Q[_V], Q[_W]] ./ Q[_ρ]`
- `viscstate.T` is the symmetric matrix `[VF[_τ11], VF[_τ12], VF[_τ13]; VF[_τ12], VF[_τ22], VF[_τ23]; VF[_τ13], VF[_τ23], VF[_τ33]]`

and the `flux._` are the corresponding matrices.

**Comment by vchuravy** [comment]
This already works (note how we only need the upper triangle, to define the matrix)

```
using LinearAlgebra
 A = @SMatrix [1 2 3;
                 0 1 2;
                 0 0 1]
sA = Symmetric(A)
3×3 Symmetric{Int64,SArray{Tuple{3,3},Int64,2,9}}:
 1  2  3
 2  1  2
 3  2  1
```

We might want to consider creating a special symmetric matrix type, that uses reduces less storage (similar to how `Tridiagonal` or `Diagonal`) , but since we only need the upper triangle to construct it, we might be fine.

**Comment by simonbyrne** [comment]
I've posted a simple implementation in one of the examples in #256. Thoughts/feedback appreciated.

**Comment by jkozdon** [comment]
This seemed to be mainly covered by the recently merged #233
